### PR TITLE
Spanner transactions

### PR DESCRIPTION
**Pull request by ChengyuanZhao**
@balopat PTAL, read-write transactions. 
the unit test is a good example of how these are used. 
This is like a real transaction where we provide a block of code and NOT just a list of mutations as in the POC.

### CODE REVIEW

**Comment by ChengyuanZhao** [comment]
@balopat read-only transactions also added now

File: spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/ReadOnlyTransactionSpannerTemplate.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+import org.springframework.cloud.gcp.data.spanner.core.convert.SpannerConverter;
+import org.springframework.cloud.gcp.data.spanner.core.mapping.SpannerDataException;
+import org.springframework.cloud.gcp.data.spanner.core.mapping.SpannerMappingContext;
+
+/**
+ * A {@link SpannerTemplate} that performs all operations in a single transaction.
+ * @author Chengyuan Zhao
+ */
+public class ReadOnlyTransactionSpannerTemplate extends SpannerTemplate {
```
Comment (balopat) [inline spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/ReadOnlyTransactionSpannerTemplate.java]: nit: shouldn't this be package private too?

File: spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/ReadOnlyTransactionSpannerTemplate.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+import com.google.cloud.spanner.ReadContext;
+import com.google.cloud.spanner.ReadOnlyTransaction;
+
+import org.springframework.cloud.gcp.data.spanner.core.convert.SpannerConverter;
+import org.springframework.cloud.gcp.data.spanner.core.mapping.SpannerDataException;
+import org.springframework.cloud.gcp.data.spanner.core.mapping.SpannerMappingContext;
+
+/**
+ * A {@link SpannerTemplate} that performs all operations in a single transaction.
+ * @author Chengyuan Zhao
```
Comment (balopat) [inline spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/ReadOnlyTransactionSpannerTemplate.java]: nit: we should clarify in the javadoc for both special templates that this is not meant to be directly instantiated by the user ever.

File: spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/ReadOnlyTransactionSpannerTemplate.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+	@Override
+	public <T> T performReadWriteTransaction(Function<SpannerOperations, T> operations) {
+		throw new SpannerDataException("A read-only transaction template cannot be used"
+				+ " to perform sub-transactions.");
+	}
+
+	@Override
+	public <T> T performReadOnlyTransaction(Function<SpannerOperations, T> operations, SpannerReadOptions readOptions) {
+		throw new SpannerDataException("A read-only transaction template cannot be used"
+				+ " to perform sub-transactions.");
```
Comment (balopat) [inline spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/ReadOnlyTransactionSpannerTemplate.java]: nit (please apply it for others too): "A read-only transaction is already under execution. Opening sub-transactions is not supported!"

File: spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/ReadOnlyTransactionSpannerTemplate.java (hunk 1 above)
Comment (ChengyuanZhao) [reply spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/ReadOnlyTransactionSpannerTemplate.java]: yes, will fix!

File: spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/ReadOnlyTransactionSpannerTemplate.java (hunk 2 above)
Comment (ChengyuanZhao) [reply spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/ReadOnlyTransactionSpannerTemplate.java]: yes will add!

File: spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/ReadOnlyTransactionSpannerTemplate.java (hunk 3 above)
Comment (ChengyuanZhao) [reply spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/ReadOnlyTransactionSpannerTemplate.java]: fixed!

**Comment by ChengyuanZhao** [comment]
@balopat changes made

**Pull request merged**